File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    boolean isBST = true;
    TreeNode prev = null;
    public void traverseInorder(TreeNode node) {
        //base Case
        if(node == null) return;
        
        //left recursive call
        traverseInorder(node.left);
        //Inorder area ---->>>>>>
        //condition
        if(prev!= null && prev.val >= node.val) isBST = false;
        //update prev
        prev = node;
        //Inorder area ---->>>>>>
        //right recursive call
        traverseInorder(node.right);
        
    }
    public boolean isValidBST(TreeNode root) {
        traverseInorder(root);
        return isBST;
    }
}

File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isBTree = true;
    TreeNode* prev = nullptr;
    void isValidBST_01(TreeNode* root) {
        if(!root) return;
        
        isValidBST_01(root->left);
        if(prev!= nullptr && prev->val >= root->val) isBTree = false;
        prev = root;
        isValidBST_01(root->right);
    }
    bool isValidBST(TreeNode* root) {
        isValidBST_01(root);
        return isBTree;
    }
};